Src/FluentAssertions/Equivalency/Node.cs

```diff
@@ -13,15 +14,35 @@ public class Node : INode
 {
     private static readonly Regex MatchFirstIndex = new(@"^\[\d+\]$");
 
+    private string path;
+    private string name;
+    private string pathAndName;
+
     public GetSubjectId GetSubjectId { get; protected set; } = () => string.Empty;
 
     public Type Type { get; protected set; }
 
-    public string Path { get; protected set; }
+    public string Path
+    {
+        get => path;
+        protected set
+        {
+            path = value;
+            pathAndName = null;
+        }
+    }
 
-    public string PathAndName => Path.Combine(Name);
+    public string PathAndName => pathAndName ??= Path.Combine(Name);
```

Comment:
Is this really that expensive?

Reply:
Expensiveness might be a relative term, as it has to compared against code maintainability.
The raw numbers in terms of performance are listed in the last benchmark.

The operation of concatenating strings (of course) depends on the length of the member name and how far down the member path we are.
The benchmarks performed have quite short type and member names, so a more realistic example with more descriptive/longer names would benefit even more from this.

Many of the calls to `PathAndName` are avoided in 8eeb20c6744eea033ed732d3125ce37501045433 as `ObjectInfo.ctor` calls `PathAndName` and also by using `Path` and `Name` separately in `Node.Equals` and `Node.GetHashCode`.

The remaining calls to `PathAndName` comes from
* `Node.Depth`
* `Node.IsRoot`
* `EquivalencyValidationContext.IsCyclicReference`
* `Property.ctor`

For this example

```cs
var subject = new MyClass[]
{
    new () { Value = 1, Inner = new () { Value = 2 } },
    new () { Value = 1, Inner = new () { Value = 2 } },
    new () { Value = 1, Inner = new () { Value = 2 } },
    new () { Value = 1, Inner = new () { Value = 2 } },
    new () { Value = 1, Inner = new () { Value = 2 } },
};
var expected = = new MyClass[]
{
    new () { Value = 1, Inner = new () { Value = 2 } },
    new () { Value = 1, Inner = new () { Value = 2 } },
    new () { Value = 1, Inner = new () { Value = 2 } },
    new () { Value = 1, Inner = new () { Value = 2 } },
    new () { Value = 1, Inner = new () { Value = 2 } },
};

subject.Should().BeEquivalentTo(expected);
```

the caching of `PathAndName` reduces the number of _allocating_ calls to `PathName` from 90 to 20.